test/Microsoft.ML.Functional.Tests/Debugging.cs

```diff
@@ -0,0 +1,214 @@
+
+            // Verify the column names.
+            int i = 0;
+            foreach (var column in data.Schema)
+            {
+                if (i == 0)
+                    Assert.Equal("Label", column.Name);
+                else
+                    Assert.Equal(HousingRegression.Features[i-1], column.Name);
+                i++;
+            }
+
+            // Verify that I can cast it to the right schema by inspecting the first row.
+            foreach (var row in mlContext.Data.CreateEnumerable<HousingRegression>(mlContext.Data.TakeRows(data, 1), true))
+            {
+                // Validate there was data in the row by checking that some values were not zero since zero is the default.
+                var rowSum = row.MedianHomeValue;
+                foreach (var property in HousingRegression.Features)
+                    rowSum += (float) row.GetType().GetProperty(property).GetValue(row, null);
+
+                Assert.NotEqual(0, rowSum);
+            }
+        }
+
+        /// <summary>
+        /// Debugging: The progress of training can be accessed.
+        /// </summary>
+        [Fact]
+        public void ViewTrainingOutput()
+        {
+            var mlContext = new MLContext(seed: 1);
+
+            // Attach a listener.
+            var logWatcher = new LogWatcher();
+            mlContext.Log += logWatcher.ObserveEvent;
+
+            // Get the dataset.
+            var data = mlContext.Data.LoadFromTextFile<HousingRegression>(GetDataPath(TestDatasets.housing.trainFilename), hasHeader: true);
+
+            // Define a pipeline
+            var pipeline = mlContext.Transforms.Concatenate("Features", HousingRegression.Features)
+                .Append(mlContext.Transforms.Normalize())
+                .AppendCacheCheckpoint(mlContext)
+                .Append(mlContext.Regression.Trainers.Sdca(
+                    new SdcaRegressionTrainer.Options { NumberOfThreads = 1, MaximumNumberOfIterations = 20 }));
+
+            // Fit the pipeline to the data.
+            var model = pipeline.Fit(data);
+
+            // Validate that we can read lines from the file.
+            var expectedLines = new string[3] {
+                @"[Source=SdcaTrainerBase; Training, Kind=Info] Auto-tuning parameters: L2 = 0.001.",
+                @"[Source=SdcaTrainerBase; Training, Kind=Info] Auto-tuning parameters: L1Threshold (L1/L2) = 0.",
+                @"[Source=SdcaTrainerBase; Training, Kind=Info] Using best model from iteration 7."};
+            foreach (var line in expectedLines)
+            {
+                Assert.Contains(line, logWatcher.Lines);
+                Assert.Equal(1, logWatcher.Lines[line]);
+            }
+        }
+
+        internal class LogWatcher {
+
+            public readonly IDictionary<string, int> Lines;
+
+            public LogWatcher()
+            {
+                Lines = new Dictionary<string, int>();
```

Comment:
>Dictionary [](start = 28, length = 10)

do we need a dictionary for this ?

Reply:
Good question. I want to hold all the lines, keep a count of how many times I see a unique set of characters, and have the lookup be quick. A hash seems like the way to go, and a `Dictionary` is a type-safe hash ([obligatory stackoverflow link](https://stackoverflow.com/questions/301371/why-is-dictionary-preferred-over-hashtable-in-c) ).

---
In reply to: [265409282](https://github.com/dotnet/machinelearning/pull/2937#discussion_r265409282) [](ancestors = 265409282)

Reply:
I'll keep as is because perf tradeoff is nominal.

---
In reply to: [265410723](https://github.com/dotnet/machinelearning/pull/2937#discussion_r265410723) [](ancestors = 265410723,265409282)